xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -633,24 +758,55 @@ private void handleCdsResponse(DiscoveryResponse cdsResponse) {
         cdsResponse.getVersionInfo());
 
     // Update local CDS cache with data in this response.
+    absentCdsResources.removeAll(clusterUpdates.keySet());
+    for (String clusterName : clusterNamesToClusterUpdates.keySet()) {
+      if (!clusterUpdates.containsKey(clusterName)) {
+        // Some previously existing resource no longer exists.
+        absentCdsResources.add(clusterName);
+      }
+    }
     clusterNamesToClusterUpdates.clear();
     clusterNamesToClusterUpdates.putAll(clusterUpdates);
 
     // Remove EDS cache entries for ClusterLoadAssignments not referenced by this CDS response.
+    for (String clusterName : clusterNamesToEndpointUpdates.keySet()) {
+      if (!edsServices.contains(clusterName)) {
+        absentEdsResources.add(clusterName);
```

Comment:
I think the right thing to do is to remove the entry from `clusterNamesToEndpointUpdates` instead of adding `absentEdsResources`. Because no new EDS watcher will subscribe to the resource.

Reply:
We need to do both in order to keep the consistency of `clusterNamesToEndpointUpdates` and `absentEdsResources`. i.e., we need both line 774 and line 787. Otherwise, if some resource name appears in both `clusterNamesToEndpointUpdates` and `absentEdsResources`, this is an invalid logic (that's why I was trying to use `clusterNamesToEndpointUpdates` cache only and use a special placeholder for known-to-be-non-existing resources to avoid maintaining this invariant).